doc/read-the-docs-site/adr/0003-sharing-prod-debugging-cek.md

```diff
@@ -0,0 +1,117 @@
+
+However, if we share code between the two machines, the performance of the production machine may be compromised.
+
+This ADR proposes an approach (originally posted [here](https://github.com/input-output-hk/plutus/pull/4909#issuecomment-1286189092)) for the two machines to share code while not compromising performance. 
+
+## Decision: Polymorphic compute/return steps
+
+As long as the debugging machine has the same type, we can alter `computeCek`/`returnCek` to be polymorphic over a type-level `Bool` specifying if we’re in debug mode or not. Then we demote it to the term level in the definition of `computeCek`/`returnCek` and branch on the `Bool` thus implementing different logic depending on whether we're in debug mode or not. This promotion to the type level allows us to statically instantiate the `Bool` in an instance and thus make GHC compile the whole worker of the CEK machine twice: once in debug mode and once in production mode. Theoretically, GHC will take care to remove all the dead debug code when in production mode.
+
+### Detailed description with code snippets
+
+Whether we are debugging or not, we will be returning a `CekState`:
+
+```haskell
+data CekState uni fun =
+    -- the next state is computing
+    Computing WordArray (Context uni fun) (Closure uni fun)
+    -- the next state is returning
+    | Returning WordArray (Context uni fun) (CekValue uni fun)
+    -- evaluation finished
+    | Terminating (Term NamedDeBruijn uni fun ())
+
+data Closure uni fun = 
+  Closure (Term NamedDeBruijn uni fun ()) (CekValEnv uni fun)
+```
+
+We enter either modes via `enterComputeCek`, which takes an extra `Bool` than our current implementation, to indicate whether we are in debugging mode or not:
+
+```haskell
+enterComputeCek 
+    :: forall uni fun s
+    . (Ix fun, PrettyUni uni fun, GivenCekReqs uni fun s)
+    => Bool
+    -> Context uni fun
+    -> Closure uni fun
+    -> CekM uni fun s (CekState uni fun)
+enterComputeCek debug ctx (Closure term env) =
+    computeCekStep (toWordArray 0) NoFrame (Closure term Env.empty) where
+    
+    computeCek
+        :: WordArray -- for costing
+        -> Context uni fun
+        -> Closure uni fun
+        -> CekM uni fun s (CekState uni fun)
+    computeCek = if debug then computeCekDebug else computeCekStep
+    {-# NOINLINE computeCek #-}  -- Making sure the `if` is only evaluated once.
+
+    -- in debugging mode, immediately returns the current `CekState` and halts execution. Debugging mode details to be worked out.
+    computeCekDebug 
+        :: WordArray
+        -> Context uni fun
+        -> Closure uni fun
+        -> CekM uni fun s (CekState uni fun) 
+    computeCekDebug budget ctx (Closure term env) = 
+        pure $ Computing budget ctx (Closure term env)
+
+    -- in production mode, recursively calls `computeCek`, and thus `computeCekStep`, similar to our current production machine.
+    computeCekStep 
```

Comment:
@thealmarty you don't need to prototype. It's just not very clear to me, and I don't think I know how to prototype this if I want to do so. My main question is how do we end up with the same code as we have now when `debug = False`.

So, I guess my confusion is because this function and `returnCekStep` are incorrectly named? Because they are used when `debug = False`, but we shouldn't compute one step at a time when not debugging.